File: kmeans.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.spatial.distance import euclidean


def k_means_clustering(X, K, n_iters):
    
    centroids_inds = []
    
    for n_k in range(K):
        ind = np.random.randint(0,len(X),1)
        centroids_inds.extend(ind)
    
    centroids = X[centroids_inds]
    
    for iteration in range(n_iters):

        # Cluster assignment step
        clusters = np.zeros((len(X)))
        for i in range(len(X)):
            distances = []
            for d in range(K):
                distance = euclidean(X[i], centroids[d])
                distances.append(distance)
            clusters[i] = distances.index(min(distances))
        
        # Move centroids step
        for k in range(K):
            centroids[k] = 1/len(clusters[clusters==k]) * np.sum(X[clusters==k], axis=0)

    return centroids, clusters



if __name__ == "__main__":

    data = pd.read_csv("C:\\Users\\Kamil\\My_repo\\OTHER-WAY\\blobs.csv").to_numpy()
    X = data[:, :-1]
    y = data[:, -1]

    plt.figure(dpi=250)
    plt.title("Raw data")
    plt.scatter(X[:,0], X[:,1], color="navy")
    plt.show()

    cents, clusts = k_means_clustering(X,4,5)

    plt.figure(dpi=250)
    plt.title("Centroids")
    plt.scatter(X[:,0], X[:,1], color="navy")
    for centroid in range(len(cents)):
        plt.scatter(cents[centroid,0], cents[centroid,1], color="red", marker="x")
    plt.show()

    plt.figure(dpi=250)
    plt.title("Clustering")
    for centroid, cluster, color in zip(range(len(cents)), range(len(cents)), ["thistle", "black", "gold", "lightblue"]):
        plt.scatter(X[:,0][clusts==cluster], X[:,1][clusts==cluster], color=color, 
                    label="cluster {}".format(cluster))
        plt.scatter(cents[centroid,0], cents[centroid,1], color="red", marker="x")
    plt.show()

File: kmeans.jl
using Pkg;
using ScikitLearn;
using PyCall;
using PyPlot;
using Distributions;
using Random;
using CSV;
using DataFrames;

Random.seed!(110)

data = CSV.read("C:\\Users\\Kamil\\My_repo\\OTHER-WAY\\blobs.csv", DataFrame);
df = Matrix(data);
X = df[:, 1:2];
y = df[:, 3];

# Starting figure
figure()
scatter(X[:,1], X[:,2], color="navy")


function k_means_clustering(X, K, n_iters)

    global centroids_ids = [];

    for n_k=1:K
        append!(centroids_ids, rand(1:size(X)[1], 1));
    end

    global centroids = X[centroids_ids,:];

    for iter=1:n_iters

        # Cluster assignment step
        global clusters = zeros(size(X)[1]);
        for i=1:size(X)[1]
            distances = []
            for d=1:K
                distance = sqrt.(sum((X[i,:] - centroids[d,:]).^2));
                #distance = euclidean(X[i,:], centroids[d,:])
                append!(distances, distance);
            end
            clusters[i] = argmin(distances)
        end

        # Move centorids step
        for k=1:K
            centroids[k,:] = (1/length(clusters[findall(==(k), clusters)])) * sum(X[findall(==(k), clusters), :], dims=1)
        end
    end
    return centroids, clusters
end


cents, clusts = k_means_clustering(X, 4, 5)


figure()
scatter(X[:,1], X[:,2], color="navy")
for centroid=1:size(cents)[1]
    scatter(cents[centroid,1], cents[centroid,2], color="red")
end


figure()
title("Clustering")
for (centroid, cluster, color) in zip(1:size(cents)[1], 1:size(cents)[1], ["thistle", "black", "gold", "lightblue"])
    scatter(X[findall(==(cluster), clusts),1], X[findall(==(cluster), clusts),2], color=color)
    scatter(cents[centroid,1], cents[centroid,2], color="red", marker="x")
end